src/cloudevents/core/v1/event.py

```diff
@@ -0,0 +1,140 @@
+            )
+
+        if attributes["id"] is None:
+            raise ValueError("Attribute 'id' must not be None")
+
+        if not isinstance(attributes["id"], str):
+            raise TypeError("Attribute 'id' must be a string")
+
+        if not isinstance(attributes["source"], str):
+            raise TypeError("Attribute 'source' must be a string")
+
+        if not isinstance(attributes["type"], str):
+            raise TypeError("Attribute 'type' must be a string")
+
+        if not isinstance(attributes["specversion"], str):
+            raise TypeError("Attribute 'specversion' must be a string")
+
+        if attributes["specversion"] != "1.0":
+            raise ValueError("Attribute 'specversion' must be '1.0'")
+
+        if "time" in attributes:
+            if not isinstance(attributes["time"], datetime):
+                raise TypeError("Attribute 'time' must be a datetime object")
+
+            if not attributes["time"].tzinfo:
+                raise ValueError("Attribute 'time' must be timezone aware")
+
+        if "subject" in attributes:
+            if not isinstance(attributes["subject"], str):
+                raise TypeError("Attribute 'subject' must be a string")
+
+            if not attributes["subject"]:
+                raise ValueError("Attribute 'subject' must not be empty")
+
+        if "datacontenttype" in attributes:
+            if not isinstance(attributes["datacontenttype"], str):
+                raise TypeError("Attribute 'datacontenttype' must be a string")
+
+            if not attributes["datacontenttype"]:
+                raise ValueError("Attribute 'datacontenttype' must not be empty")
+
+        if "dataschema" in attributes:
+            if not isinstance(attributes["dataschema"], str):
+                raise TypeError("Attribute 'dataschema' must be a string")
+
+            if not attributes["dataschema"]:
+                raise ValueError("Attribute 'dataschema' must not be empty")
+
+        for custom_extension in (
+            set(attributes.keys()) - REQUIRED_ATTRIBUTES - OPTIONAL_ATTRIBUTES
+        ):
+            if custom_extension == "data":
+                raise ValueError(
+                    "Extension attribute 'data' is reserved and must not be used"
+                )
+
+            if not custom_extension[0].isalpha():
+                raise ValueError(
+                    f"Extension attribute '{custom_extension}' should start with a letter"
+                )
+
+            if not (5 <= len(custom_extension) <= 20):
+                raise ValueError(
+                    f"Extension attribute '{custom_extension}' should be between 5 and 20 characters long"
+                )
+
+            if not re.match(r"^[a-z0-9]+$", custom_extension):
+                raise ValueError(
+                    f"Extension attribute '{custom_extension}' should only contain lowercase letters and numbers"
+                )
```

Comment:
it's probably a good idea to:
1. split this method into multiple smaller ones with distinct validations.
2. maybe extract some reusable logic
3. maybe combine checks in order to avoid reporting attribute errors one-by-one. Meaning that we should go over all attributes and emit a combined error with all the possible attribute errors. Just as you did with required attributes

Reply:
My goal is to have this validation logic in a single place, it's easier to follow IMO and read it, rather than navigate into multiple small methods. 

About extracting reusable logic, I can try to look into this, but I also wanted to have it "match" the spec and treat each attribute as individual. Again, for readability reasons mainly. It's easier to reason about when the whole validation logic is colocated. 

> maybe combine checks in order to avoid reporting attribute errors one-by-one.

I'm a fan of "fail fast" approach, what would be an advantage of raising all exceptions at once?

Reply:
What do you think about something like this?
```
        CloudEvent._validate_required_attributes(attributes)
        CloudEvent._validate_attribute_types(attributes)
        CloudEvent._validate_optional_attributes(attributes)
        CloudEvent._validate_extension_attributes(attributes)
 ```
 and each method will hold the necesary validation logic?

Reply:
Grouping CloudEvents (CE) attributes validation with `required` and `optional` as well as having a separate `extension` rules is IMO a good idea.

On the fail fast approach - sure, that's a valid one, but IMO not in this case. I guess people would only see the validation errors when they are developing against CE, not during active usage (hopefully). So it makes sense to gather all the validation errors all together and present them together to ease analysis of a defective event. I'd probably even go and create a custom error instead of using standard ones with more details on what and where has failed the validation.